{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "cb88d65a",
   "metadata": {},
   "source": [
    "# Topics In Bioinformatics - Wk4 Assignment - Dr. Jody Hey \n",
    "\n",
    "## Madison Dautle\n",
    "## Created on Sept. 22, 2025\n",
    "## Last updated on Sept. 22, 2025"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "ab784d2d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd \n",
    "import time\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "e56cc1c0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Population Choices: \n",
      "\n",
      "['GBR' 'FIN' 'CHS' 'PUR' 'CDX' 'CLM' 'IBS' 'PEL' 'PJL' 'KHV' 'ACB' 'GWD'\n",
      " 'ESN' 'BEB' 'MSL' 'STU' 'ITU' 'CEU' 'YRI' 'CHB' 'JPT' 'LWK' 'ASW' 'MXL'\n",
      " 'TSI' 'GIH']\n"
     ]
    }
   ],
   "source": [
    "#Read in the individuals list and show population options\n",
    "individuals_df = pd.read_csv('./integrated_call_samples_v3.20130502.ALL.panel', sep='\\t', header = 0, usecols = [0,1,2]) \n",
    "\n",
    "#unique_superpop = individuals_df['super_pop'].unique()\n",
    "unique_pop = individuals_df['pop'].unique()\n",
    "\n",
    "#print('Super Population Choices: \\n') \n",
    "#print(unique_superpop) \n",
    "print('Population Choices: \\n') \n",
    "print(unique_pop)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "c12b70de-99be-4aad-8c42-93d14203e96a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Choose (a) population(s) for analysis \n",
    "selc_pop = ['YRI','GBR','JPT'] "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "ad1e7361-f719-4f2d-99a1-41d3552a0568",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Set up output to pdf\n",
    "from reportlab.lib import colors\n",
    "from reportlab.lib.pagesizes import letter, portrait\n",
    "from reportlab.platypus import SimpleDocTemplate, Table, TableStyle, Paragraph, Spacer, PageBreak\n",
    "from reportlab.lib.styles import getSampleStyleSheet\n",
    "\n",
    "outFile = 'summary.pdf' \n",
    "pdf = SimpleDocTemplate(outFile, pagesize=portrait(letter))\n",
    "elements = []\n",
    "styles = getSampleStyleSheet() \n",
    "header_style = styles[\"Heading2\"] \n",
    "\n",
    "def write_df_to_pdf(df): \n",
    "    data = [df.columns.to_list()] + df.values.tolist()\n",
    "    table = Table(data) \n",
    "    table.setStyle(TableStyle([\n",
    "        (\"BACKGROUND\", (0, 0), (-1, 0), colors.grey),\n",
    "        (\"TEXTCOLOR\", (0, 0), (-1, 0), colors.whitesmoke),\n",
    "        (\"ALIGN\", (0, 0), (-1, -1), \"CENTER\"),\n",
    "        (\"FONTNAME\", (0, 0), (-1, 0), \"Helvetica-Bold\"),\n",
    "        (\"BOTTOMPADDING\", (0, 0), (-1, 0), 12), \n",
    "        (\"BACKGROUND\", (0, 1), (-1, -1), colors.beige), \n",
    "        (\"GRID\", (0, 0), (-1, -1), 1, colors.black), \n",
    "    ]))\n",
    "    return table\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "f325dcf5-ce74-4b26-ab20-cbf84ac8e7a1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running analysis for YRI population\n",
      "Total time to read in file: 23.891176223754883\n",
      "Total time to remove indel postions from file: 1.9350988864898682 sec\n",
      "Total time to remove indel postions from file: 32.039238691329956 sec\n",
      "Total time to remove non SNP postions from file: 5.272151231765747 sec\n",
      "Total time to count observed heterozygtes @ each SNP & calculate heterozygosity: 16.98350429534912 sec\n",
      "Total time to count allelic frequencies @ each SNP & calculate expected heterozygosity: 58.562153339385986 sec\n",
      "         Observed_heterozygosity  Expected_heterozygosity   obs-exp\n",
      "1                       0.092593                 0.088306  0.004287\n",
      "2                       0.046296                 0.062714 -0.016418\n",
      "17                      0.037037                 0.036351  0.000686\n",
      "26                      0.037037                 0.036351  0.000686\n",
      "54                      0.009259                 0.009216  0.000043\n",
      "...                          ...                      ...       ...\n",
      "1103536                 0.203704                 0.335219 -0.131516\n",
      "1103537                 0.018519                 0.018347  0.000171\n",
      "1103538                 0.018519                 0.018347  0.000171\n",
      "1103539                 0.027778                 0.027392  0.000386\n",
      "1103543                 0.064815                 0.062714  0.002100\n",
      "\n",
      "[281352 rows x 3 columns]\n",
      "Total processing time for YRI: 2.3114632328351337 min\n",
      "Running analysis for GBR population\n",
      "Total time to read in file: 20.88079595565796\n",
      "Total time to remove indel postions from file: 0.8905487060546875 sec\n",
      "Total time to remove indel postions from file: 30.486034870147705 sec\n",
      "Total time to remove non SNP postions from file: 3.4087955951690674 sec\n",
      "Total time to count observed heterozygtes @ each SNP & calculate heterozygosity: 9.920357704162598 sec\n",
      "Total time to count allelic frequencies @ each SNP & calculate expected heterozygosity: 32.64432120323181 sec\n",
      "         Observed_heterozygosity  Expected_heterozygosity   obs-exp\n",
      "6                       0.010989                 0.010929  0.000060\n",
      "18                      0.032967                 0.032424  0.000543\n",
      "33                      0.153846                 0.142012  0.011834\n",
      "56                      0.219780                 0.195629  0.024152\n",
      "69                      0.032967                 0.032424  0.000543\n",
      "...                          ...                      ...       ...\n",
      "1103527                 0.010989                 0.010929  0.000060\n",
      "1103535                 0.010989                 0.010929  0.000060\n",
      "1103536                 0.043956                 0.042990  0.000966\n",
      "1103543                 0.032967                 0.032424  0.000543\n",
      "1103546                 0.010989                 0.010929  0.000060\n",
      "\n",
      "[164173 rows x 3 columns]\n",
      "Total processing time for GBR: 1.637299859523773 min\n",
      "Running analysis for JPT population\n",
      "Total time to read in file: 21.83976435661316\n",
      "Total time to remove indel postions from file: 0.9932651519775391 sec\n",
      "Total time to remove indel postions from file: 31.709237337112427 sec\n",
      "Total time to remove non SNP postions from file: 3.8986990451812744 sec\n",
      "Total time to count observed heterozygtes @ each SNP & calculate heterozygosity: 9.617888689041138 sec\n",
      "Total time to count allelic frequencies @ each SNP & calculate expected heterozygosity: 32.45084524154663 sec\n",
      "         Observed_heterozygosity  Expected_heterozygosity   obs-exp\n",
      "22                      0.009615                 0.009569  0.000046\n",
      "33                      0.163462                 0.166004 -0.002543\n",
      "35                      0.009615                 0.009569  0.000046\n",
      "36                      0.009615                 0.009569  0.000046\n",
      "51                      0.019231                 0.019046  0.000185\n",
      "...                          ...                      ...       ...\n",
      "1103532                 0.173077                 0.158099  0.014978\n",
      "1103533                 0.009615                 0.009569  0.000046\n",
      "1103535                 0.009615                 0.009569  0.000046\n",
      "1103536                 0.240385                 0.253652 -0.013267\n",
      "1103543                 0.019231                 0.019046  0.000185\n",
      "\n",
      "[158462 rows x 3 columns]\n",
      "Total processing time for JPT: 1.6752223730087281 min\n"
     ]
    }
   ],
   "source": [
    "if not os.path.exists('./madison.vcf'):\n",
    "    print('Removing header lines from VCF file, so it can be read in with the column names intact') \n",
    "    command = \"grep -v '^##' ./ALL.chr22.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf | sed 's/^#//g' > ./madison.vcf\"\n",
    "    os.system(command)\n",
    "\n",
    "all_results = pd.DataFrame(columns=['POS', 'REF', 'ALT','Observed_heterozygosity','Expected_heterozygosity','obs-exp','Population']) \n",
    "summary_table = pd.DataFrame(index = selc_pop, columns=['Avg_obs_hetero','Avg_exp_hetero','Avg_diff'])\n",
    "for pop in selc_pop: \n",
    "    print(f'Running analysis for {pop} population')\n",
    "    start_tot_time = time.time()\n",
    "    start_time = time.time()\n",
    "    keepcols = ['POS','REF','ALT']\n",
    "    pop_samples = individuals_df.loc[individuals_df['pop'] == pop, 'sample'].tolist()\n",
    "    keepcols = keepcols+pop_samples\n",
    "    #print(keepcols)\n",
    "    vcf = pd.read_csv('./madison.vcf', sep = '\\t', usecols = keepcols, header = 0)\n",
    "    end_time = time.time() \n",
    "    print(f'Total time to read in file: {end_time-start_time}')\n",
    "    #print(vcf.columns)\n",
    "\n",
    "    \n",
    "    #Remove indels \n",
    "    start_time = time.time()\n",
    "    vcf = vcf[(vcf['REF'].str.len() == 1) & (vcf['ALT'].str.len() == 1)] \n",
    "    end_time = time.time() \n",
    "    print(f'Total time to remove indel postions from file: {end_time-start_time} sec')\n",
    "\n",
    "\n",
    "    #Remove SNPs that have more than one variant \n",
    "    start_time = time.time()\n",
    "    vcf = vcf[vcf[pop_samples].apply(lambda row: row.isin(['0|0','0|1','1|0','1|1']).all(), axis=1)]\n",
    "    end_time = time.time() \n",
    "    print(f'Total time to remove indel postions from file: {end_time-start_time} sec')\n",
    "    #print(vcf.head()) \n",
    "\n",
    "\n",
    "    #Remove positions where there are no SNPs \n",
    "    start_time = time.time()\n",
    "    mask = ~(vcf[pop_samples] == '0|0').all(axis=1)\n",
    "    vcf = vcf[mask] \n",
    "    end_time = time.time() \n",
    "    print(f'Total time to remove non SNP postions from file: {end_time-start_time} sec')\n",
    "    #print(vcf.head()) \n",
    "\n",
    "\n",
    "    #Count heterozygotes & calculate obsrved heterozygosity \n",
    "    start_time = time.time()\n",
    "    vcf['num_hets']  = vcf[pop_samples].apply(lambda row: ((row == '0|1') | (row == '1|0')).sum(), axis=1)\n",
    "    vcf['Observed_heterozygosity'] = vcf['num_hets']/len(pop_samples) \n",
    "    end_time = time.time() \n",
    "    print(f'Total time to count observed heterozygtes @ each SNP & calculate heterozygosity: {end_time-start_time} sec')\n",
    "    #print(vcf[['num_hets','Observed_heterozygosity']])\n",
    "\n",
    "\n",
    "    #Calculate allelic frequencies & expected heterozygosity \n",
    "    start_time = time.time()\n",
    "    genotype_data = vcf[pop_samples] \n",
    "    vcf['raw_num_p'] = genotype_data.apply(lambda row: row.str.replace('|','').str.count('0').sum(), axis=1)\n",
    "    vcf['raw_num_q'] = genotype_data.apply(lambda row: row.str.replace('|','').str.count('1').sum(), axis=1)\n",
    "    #vcf['test'] = len(pop_samples) - ((vcf['raw_num_p'] + vcf['raw_num_q'])/2) # all columns should be equal to 0\n",
    "    #print(vcf[['test']])\n",
    "    vcf['p'] = vcf['raw_num_p']/(vcf['raw_num_p'] + vcf['raw_num_q'])\n",
    "    vcf['q'] = vcf['raw_num_q']/(vcf['raw_num_p'] + vcf['raw_num_q'])\n",
    "    #vcf['test1'] = vcf['p'] + vcf['q'] \n",
    "    #print(vcf[['test1']]) #all columns should be equal to 1\n",
    "    vcf['Expected_heterozygosity'] = 2*vcf['p']*vcf['q']\n",
    "    end_time = time.time() \n",
    "    print(f'Total time to count allelic frequencies @ each SNP & calculate expected heterozygosity: {end_time-start_time} sec')\n",
    "    #print(vcf[['Observed_heterozygosity','Expected_heterozygosity']])\n",
    "\n",
    "\n",
    "    #Calculate Obs-Expected heterozygosity \n",
    "    vcf['obs-exp'] = vcf['Observed_heterozygosity'] - vcf['Expected_heterozygosity'] \n",
    "    print(vcf[['Observed_heterozygosity','Expected_heterozygosity','obs-exp']])\n",
    "\n",
    "\n",
    "    #Calculate average observed expected, & difference for the population and write to summary table \n",
    "    summary_table.loc[pop, 'Avg_obs_hetero'] = vcf['Observed_heterozygosity'].mean()\n",
    "    summary_table.loc[pop, 'Avg_exp_hetero'] = vcf['Expected_heterozygosity'].mean()\n",
    "    summary_table.loc[pop, 'Avg_diff'] = vcf['obs-exp'].mean()\n",
    "    summary_table = summary_table.round(5) \n",
    "    end_tot_time = time.time() \n",
    "\n",
    "\n",
    "    #Add population dataframe to pdf \n",
    "    pop_df = vcf[['POS', 'REF', 'ALT','Observed_heterozygosity','Expected_heterozygosity','obs-exp']].copy()\n",
    "    pop_df = pop_df.round(5)\n",
    "    elements.append(Paragraph(f\"Population: {pop}\", header_style))\n",
    "    elements.append(Spacer(1, 12))\n",
    "    elements.append(write_df_to_pdf(pop_df))\n",
    "    elements.append(PageBreak())\n",
    "    del pop_df\n",
    "    del vcf\n",
    "    \n",
    "    print(f'Total processing time for {pop}: {(end_tot_time - start_tot_time)/60} min') \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "3b2eaa26-ce74-4a1a-aba4-1d11b031812b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total time taken to write summary of results to pdf: 1854.6915848255157 sec\n"
     ]
    }
   ],
   "source": [
    "#Add summary table to pdf\n",
    "start_time = time.time()\n",
    "\n",
    "elements.append(Paragraph(\"Summary Table\", header_style))\n",
    "elements.append(Spacer(1, 12))\n",
    "summary_tbl = write_df_to_pdf(summary_table.reset_index())\n",
    "elements.append(summary_tbl)\n",
    "elements.append(PageBreak())\n",
    "\n",
    "#Write pdf to file\n",
    "pdf.build(elements)\n",
    "\n",
    "end_time = time.time()\n",
    "\n",
    "print(f'Total time taken to write summary of results to pdf: {(end_time-start_time)} sec') "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "66e1c2c5-357f-42cc-9715-a72dcffda272",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
